napari/components/layerlist.py

```diff
@@ -59,21 +59,43 @@ def __init__(self, data=()) -> None:
             basetype=Layer,
             lookup={str: lambda e: e.name},
         )
+        self._create_contexts()
+
+    def _create_contexts(self):
+        """Create contexts to manage enabled/visible action/menu states.
+
+        Connects LayerList and Selection[Layer] to their context keys to allow
+        actions and menu items (in the GUI) to be dynamically enabled/disabled
+        and visible/hidden based on the state of layers in the list.
+        """
 
         # TODO: figure out how to move this context creation bit.
         # Ideally, the app should be aware of the layerlist, but not vice versa.
-        # This could probably be done by having the layerlist emit events that the app
-        # connects to, then the `_ctx` object would live on the app, (not here)
+        # This could probably be done by having the layerlist emit events that
+        # the app connects to, then the `_ctx` object would live on the app,
+        # (not here)
         from napari._app_model.context import create_context
         from napari._app_model.context._layerlist_context import (
             LayerListContextKeys,
+            LayerListSelectionContextKeys,
         )
 
         self._ctx = create_context(self)
         if self._ctx is not None:  # happens during Viewer type creation
             self._ctx_keys = LayerListContextKeys(self._ctx)
-
-            self.selection.events.changed.connect(self._ctx_keys.update)
+            self.events.inserted.connect(self._ctx_keys.update)
+            self.events.removed.connect(self._ctx_keys.update)
+
+        self._selection_ctx = create_context(self)
```

Comment:
@tlambert03 this may be a silly question but when I create another `Context` using `self` here, would the `id` of `self` be the same? So when [`create_context`](https://github.com/pyapp-kit/app-model/blob/4d796c6390c2c4141f370f93a3162731e0d940e5/src/app_model/expressions/_context.py#L80) does:

```python
_OBJ_TO_CONTEXT[obj_id] = new_context
```

am I replacing the old dict item?

Reply:
nope, each `LayerList` instance will have a unique `id()`; more generally, all python objects (that haven't been garbage collected)  have a unique id... anytime `object_a is not object_b`, it means they have a different id.

Reply:
Sorry I should have been explicit in that I use `create_context` twice in the init, i.e. in the same instance. But thinking about it I think instead of:

```python
        self._ctx = create_context(self)
        if self._ctx is not None:  # happens during Viewer type creation
            self._ctx_keys = LayerListContextKeys(self._ctx)
            self.events.inserted.connect(self._ctx_keys.update)
            self.events.removed.connect(self._ctx_keys.update)

        self._selection_ctx = create_context(self)
        if (
            self._selection_ctx is not None
        ):  # happens during Viewer type creation
            self._selection_ctx_keys = LayerListSelectionContextKeys(
                self._selection_ctx
            )
```

it should just be:

```python
        self._ctx = create_context(self)
        if self._ctx is not None:  # happens during Viewer type creation
            self._ctx_keys = LayerListContextKeys(self._ctx)
            self.events.inserted.connect(self._ctx_keys.update)
            self.events.removed.connect(self._ctx_keys.update)

            self._selection_ctx_keys = LayerListSelectionContextKeys(
                self._ctx
            )
```

i.e. I can just use the same context to create the `LayerListContextKeys` and `LayerListSelectionContextKeys`.

Is that right?

Reply:
> new context for the same object... why?

I didn't know what I was doing 😅

Reply:
Aw, well, no biggie.  Are you calling get_context anywhere on the layerlist? If not, no harm done, and easily changed

Reply:
>  Are you calling get_context anywhere on the layerlist?

Yes I do, I manually up `update_from_context` for all the menubar menus:

https://github.com/napari/napari/blob/98c2e0f39982d46aebc70acf5a32dd2beaa23df4/napari/_qt/qt_main_window.py#L817

which is used here:

https://github.com/napari/napari/blob/98c2e0f39982d46aebc70acf5a32dd2beaa23df4/napari/_qt/qt_main_window.py#L868-L874

I thought I checked and this still worked...?

Does my double up in calling `get_context` create a different `Context` the second time round? Or is the double up just redundant but doesn't change anything?

Reply:
ok... looked at this a bit more.   It does work as you have it here, but I'd say the fact that it works is a bit surprising, and probably shouldn't be depended on as a pattern.  It works because when you call `self._selection_ctx = create_context(self)` (after having already created a context for `self`), you end up hitting [this branch in app_model's `create_context`](https://github.com/pyapp-kit/app-model/blob/81b245baa45cb3bc7e1afd81d3af06dfe6fa1ca9/src/app_model/expressions/_context.py#L134-L140):

```python
            # FIXME: this might be a bit napari "magic"
            # it also assumes someone uses "self" as the first argument
            if "self" in frame.f_locals:
                _ctx = _OBJ_TO_CONTEXT.get(id(frame.f_locals["self"]))
                if _ctx is not None:
                    parent = _ctx
```

What that chunk does is it asks: "have we already created a context for this object? if so, then scope the context we are about to create off of this existing context"

Meaning: the new context you get back `self._selection_ctx` is a `ChainMap` that has the previously created context `self._ctx`.  (and so any keys in the original self._ctx can also be accessed from self._selection_ctx) You can see this by running:

```python
        self._selection_ctx = create_context(self)
        print("has self._ctx?:", self._ctx in self._selection_ctx.maps)
```

(If that weren't the case, then nothing would be "responding" to changes in the `LayerListContextKeys.num_layers`, which, at the moment, appears to only be the `napari.window.file.save_layers_dialog` Action... so, without this chain map magic here, your `Save All Layers...` menu item would never be enabled).

in summary, it's working fine, but probably only by accident.

Reply:
Thanks, Draga and I had a dig and thought the same!

I guess it would be hard to warn if a context has been created with the same object id before without a context registry as you'd need to keep track of ids.

Reply:
> I guess it would be hard to warn if a context has been created with the same object id before without a context registry as you'd need to keep track of ids.

I mean we track it in `_OBJ_TO_CONTEXT` so we could just check if the id is already a key in the dictionary, no?

Reply:
:facepalm: omg yes of course, would be trivial.